File: csharp/TestUsernamePasswordSessionHelper.cs
 /**
  * 	License: www.workbooks.com/mit_license
  * 	Last commit $Id$
  *
  */
using System;
using System.Collections.Generic;
using WorkbooksApiApplication;

namespace WorkbooksApiApplication
{
  public class TestUsernamePasswordSessionHelper
  {
    Dictionary<string, object> loginParams = new Dictionary<string, object>();
    string service = "http://localhost:3000";
    string application_name = "csharp_test_client";
    string user_agent = "csharp_test_client/0.1";
    string username = "demo@workbooks.com";
    string password = "demo1234";
    bool verify_peer = false;

    public WorkbooksApi workbooks = null;

    public TestUsernamePasswordSessionHelper ()
    {
      if (service != null) {
        loginParams.Add("service", service);
      }
      loginParams.Add("application_name", application_name);
      loginParams.Add("user_agent", user_agent);
      loginParams.Add("verify_peer", verify_peer);
      try {
        workbooks = new WorkbooksApi(loginParams);
      } catch(Exception e) {
        Console.WriteLine ("Exception while getting the workbooks object" + e.Message);
      }
    }

    public WorkbooksApi testLogin() {
      loginParams = new Dictionary<string, object>();
      loginParams.Add("username", username);
      loginParams.Add("password", password);
      if (System.Environment.GetEnvironmentVariable("DATABASE_ID") != null) {
        loginParams.Add("logical_database_id", System.Environment.GetEnvironmentVariable("DATABASE_ID"));
      }
      workbooks.log("Login commences", new Object[] {this.GetType().Name});
      Dictionary <string, object> response = null;
      try {
        response = workbooks.login(loginParams);
        int http_status = (int) response["http_status"];
        Dictionary<string, object> responseData = (Dictionary<string, object>) response["response"];
        if (response.ContainsKey("failure_reason")) {
          String failure_reason = response["failure_reason"].ToString();
          if (http_status == WorkbooksApi.HTTP_STATUS_FORBIDDEN && failure_reason.Equals("no_database_selection_made")) {
            String default_database_id = responseData["default_database_id"].ToString();
            loginParams.Add("logical_database_id", default_database_id);

            response = workbooks.login(loginParams);
            http_status = (int)response["http_status"];
            if (http_status != WorkbooksApi.HTTP_STATUS_OK) {
              workbooks.log("Login has failed", new Object[] {response}, "error");
              System.Environment.Exit(0);
            }
          }
        }
        workbooks.log("Login complete", new Object[] {this.GetType().ToString()}, "info");
      } catch(Exception e) {      
        workbooks.log("Exception: ", new Object[] {e}, "error");
      }
      return workbooks;
    }

    public void testExit(WorkbooksApi workbooks, int exitcode) {
      try {
        Dictionary<string, object> logout = workbooks.logout();
        if (!logout.ContainsKey("success") ) {
          workbooks.log("Logout failed", new Object[] {logout}, "error");
          System.Environment.Exit(0);
        }
        workbooks.log("Logout Complete", new Object[] {this.GetType()}, "info");
      } catch(Exception e) {
        workbooks.log("Errors while logging out: ", new Object[] {e}, "error");
      }
      if (exitcode == 0) {
        workbooks.log("Script exited", new Object[] {"OK"}, "info");
      } else {
        workbooks.log("Script exited with error", new Object[] {exitcode}, "error");
      }
      System.Environment.Exit(exitcode);
    }

  } // class
}

File: java/TestUsernamePasswordSessionHelper.java
package workbooks_app.client_lib.java;
import java.util.HashMap;
import javax.json.JsonObject;

/** 
 *	Login wrapper for Workbooks for API test purposes. This version uses a 'traditional' username and 
 *  password. As an alternative, consider using an API Key without explicit Login/Logout requests: 
 *  see TestLoginHelper.java
 * 
 * 	License: www.workbooks.com/mit_license
 * 	Last commit $Id: TestUsernamePasswordSessionHelper.java 22080 2014-05-21 12:53:52Z bviroja $
 */
 
public class TestUsernamePasswordSessionHelper {

	
	HashMap<String, Object> params = new HashMap<String,Object>();
	String service = "http://localhost:3000";
	String application_name = "java_test_client";
	String user_agent = "java_test_client/0.1";
	String username = "demo@workbooks.com";
	String password = "demo1234";
	boolean verify_peer = false;
	
	WorkbooksApi workbooks = null;
	
	public TestUsernamePasswordSessionHelper() {
		if (service != null) {
			params.put("service", service);
		}
		params.put("application_name", application_name);
		params.put("user_agent", user_agent);
		params.put("logger_callback", "logAllToStdout");
		params.put("verify_peer", verify_peer);

		try {
			workbooks = new WorkbooksApi(params);
		} catch(Exception e) {
			
		}
	}
	
	public WorkbooksApi testLogin() {
		HashMap<String, Object> loginParams = new HashMap<String, Object>();
		loginParams.put("username", username);
		loginParams.put("password", password);
		
		if (System.getenv("DATABASE_ID") != null) {
			loginParams.put("logical_database_id", System.getenv("DATABASE_ID"));
		}
		workbooks.log("Login commences", new Object[] {this.getClass().toString()});
		HashMap<String, Object> response = null;
		try {
			response = workbooks.login(loginParams);
			int http_status = (Integer) response.get("http_status");
			//int http_status = Integer.parseInt((String) response.get("http_status"));
	
			JsonObject responseObject = (JsonObject)response.get("response"); 
			if (response.containsKey("failure_reason")) {
				String failure_reason = response.get("failure_reason").toString();
				if (http_status == WorkbooksApi.HTTP_STATUS_FORBIDDEN && failure_reason.equals("no_database_selection_made")) {
					String default_database_id = responseObject.getString("default_database_id");
					loginParams.put("logical_database_id", default_database_id);
					
					response = workbooks.login(loginParams);
					http_status = (Integer)response.get("http_status");
					if (http_status != WorkbooksApi.HTTP_STATUS_OK) {
						workbooks.log("Login has failed", new Object[] {response}, "error", WorkbooksApi.DEFAULT_LOG_LIMIT);
						System.exit(0);
					}
				}
			}
			
			workbooks.log("Login complete", new Object[] {this.getClass()}, "info", WorkbooksApi.DEFAULT_LOG_LIMIT);
			
		} catch(Exception e) {			
			workbooks.log("Exception: ", new Object[] {e}, "error", WorkbooksApi.DEFAULT_LOG_LIMIT);
		}
		
		return workbooks;
	}
		
	public void testExit(WorkbooksApi workbooks, int exitcode) {
		try {
			HashMap<String, Object> logout = workbooks.logout();
			if (!logout.containsKey("success") ) {
				workbooks.log("Logout failed", new Object[] {logout}, "error", WorkbooksApi.DEFAULT_LOG_LIMIT);
				System.exit(0);
			}
			workbooks.log("Logout Complete", new Object[] {this.getClass()}, "info", WorkbooksApi.DEFAULT_LOG_LIMIT);
		} catch(Exception e) {
			workbooks.log("Errors while logging out: ", new Object[] {e}, "error", WorkbooksApi.DEFAULT_LOG_LIMIT);
		}
		
		if (exitcode == 0) {
			workbooks.log("Script exited", new Object[] {"OK"}, "info", WorkbooksApi.DEFAULT_LOG_LIMIT);
		} else {
			workbooks.log("Script exited with error", new Object[] {Integer.toString(exitcode)}, "error", WorkbooksApi.DEFAULT_LOG_LIMIT);
		}
		System.exit(exitcode);
		
	}
	
	public static void main(String[] args) {
		TestUsernamePasswordSessionHelper testApi = new TestUsernamePasswordSessionHelper();
		//testApi.testLogin();
		testApi.testJsonConversion();
	}
		
}
